docs/dev/typesupport.md

```diff
@@ -0,0 +1,185 @@
+representation of the TypeTree. But, the output and the internal flow vary
+greatly between languages.
+
+For instance, C++ requires only a header and a source file. Java, however,
+requires each (public) class to reside in a separate file.
+
+Previous incarnations of IDL compilers studied often used a visitor pattern.
+i.e. The backend registers a set of callback functions, usually one or more
+for each type, and the programmer calls a function that *visits* each type
+exactly once and calls the appropriate callback functions in order. However,
+this method is considered to rigid.
+
+1. There are cases where a part of the tree must be traversed several times
+   with different callback functions. For instance, when generating code for
+   for a struct in an object oriented language, all members must be traversed
+   before the class definition, constructors and destructors can be generated.
+
+2. There could be a need to adjust the order in which the tree is traversed.
+   For instance when a given language does not support nested structs. The
+   order in which the tree is traversed must be reversed to first emit the
+   definitions of the nested structs before the definition of the current
+   struct can be emitted.
+
+3. ...
+
+One of the design goals of the intermediate format and accompanying utilities
+is to facilitate efficient development of backends. Therefore, because there
+is no single pattern that works well for all supported languages, the TypeTree
+itself will be considered the API and basic utility functions will be offered
+to simplify traversing the TypeTree etc.
+
+
+### TypeTree structure
+
+The TypeTree, as indicated by the name, is a simple tree that closely follows
+IDL syntax. This is a requirement because the order has importance. A
+pseudocode example of how types are constructed is provided below.
+
+```C
+#define INT (1<<0)
+// ...
+#define TYPEDEF (1<<10)
+#define STRUCT (1<<11)
+#define MAP (1<<12)
+// ...
+#define ARRAY (1<<20)
+#define UNBOUND (1<<21)
+
+typedef union type type_t;
+
+typedef void(*dtor_t)(type_t *type, type_t *parent);
+
+typedef struct {
+  int flags;
+  char *name;
+  type_t *next;
+  type_t *parent;
+  dtor_t dtor;
+} typespec_t;
+
+typedef struct {
+  typespec_t type; /* ARRAY bit set icw. normal type flags and perhaps UNBOUND */
+  size_t size;
+} array_t;
+
+typedef struct { /* TYPEDEF bit set, name is name of typedef. */
+  typespec_t type;
+  typespec_t *target; /* Pointer to target type. */
+} typedef_t;
+
+typedef struct {
+  typespec_t type; /* FORWARD_DECL and type bits set. */
+  type_t *target;
+} forward_decl_t;
+
+/* Constructed types introduce a complexity when the TypeTree is destructed
+   because they can be both embedded and referenced. A constructed type must
+   only be destructed from the scope in which it is defined, never when merely
+   referenced. */
+typedef struct {
+  typespec_t type;
+  type_t *members;
```

Comment:
It is important to note that the members are heterogeneous, in the sense that the elements may have different types, just like a module can contain different type of members, such struct and union definitions.

There is a certain overlap between what can occur in a module and a struct. If one would define a walk function that walks over different types in the type tree, it would be handy if somehow the members are shared, so that in the implementation you do not need to differentiate between them.